website/content/en/highlights/2022-03-15-vrl-vm-beta.md

```diff
@@ -0,0 +1,122 @@
+---
+date: "2022-03-15"
+title: "Beta release of a Virtual Machine to run VRL programs"
+description: "A faster way to run VRL"
+authors: ["StephenWakely"]
+pr_numbers: [9829]
+release: "0.21.0"
+hide_on_release_notes: false
+badges:
+  type: enhancement
+---
+
+As part of our performance improvements in Vector, we have developed a new Virtual Machine for
+running VRL programs.
+
+In our soak tests we are getting performance improvements of between 10-15% depending on
+the VRL that is being run.
+
+## Opting in
+
+You can give the VM a try by adding the `runtime` configuration option to your Remap transform:
+
+```toml
+[transforms.remap]
+type = "remap"
+inputs = [ "..." ]
+runtime = "vm"
+source = '''
+...
+'''
+```
+
+## Let us know what you think!
+
+We are excited about the performance improvements that using a Virtual Machine will bring to Vector.
+If you have any feedback for us let us know on [Discord] or on [Twitter].
+
+## Under the hood
+
+Currently VRL is run using a Tree Walking interpreter. A VRL program is compiled into a tree of
+program nodes. For example, given this simple VRL program:
+
+```coffee
+if .flag == 42 {
+  .field = 32
+} else {
+  .field = 15
+}
+```
+
+VRL compiles that into the nodes similar to this:
+
+```mermaid
+graph TD
+    IF -->PREDICATE
+    PREDICATE -->Equals[==]
+    Equals -->Flag[.flag]
+    Equals -->42
+
+    IF -->THEN
+    THEN -->ASSIGN1[ASSIGN]
+    ASSIGN1 -->Field1[.field]
+    ASSIGN1 -->32
+
+    IF -->ELSE
+    ELSE -->ASSIGN2[ASSIGN]
+    ASSIGN2 -->Field2[.field]
+    ASSIGN2 -->15
+
+```
+
+As the VRL program is run each node is interpreted which decides what actions need to be
+taken and which node to execute next. The tree is walked in this manner until the program
+reaches a node that has no further nodes to execute.
+
+There are a number of advantages to running a program this way. It is fast to develop and the
+code is easy to maintain and debug since each node manages it's own state. Also, especially
+since it is developed in Rust, the code is quite safe with little opportunity for undefined
+behaviour to creep in.
```

Comment:
As a reader, this statement would concern me, especially since there's no other mention in this highlight on how much _less_ safe enabling the VM is for me as a user.

Reply:
Yeah, agreed, I think we can just drop this statement.